L1Trigger/Phase2L1ParticleFlow/plugins/L1HPSPFTauProducer.cxx

```diff
@@ -0,0 +1,204 @@
+#include <vector>
+#include <numeric>
+
+////////////////////
+// FRAMEWORK HEADERS
+#include "FWCore/Framework/interface/global/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#include "DataFormats/L1TParticleFlow/interface/PFCandidate.h"
+#include "DataFormats/L1TParticleFlow/interface/PFTau.h"
+#include "DataFormats/JetReco/interface/CaloJet.h"
+// bitwise emulation headers
+#include "L1Trigger/Phase2L1ParticleFlow/interface/taus/L1HPSPFTauEmulator.h"
+
+
+class L1HPSPFTauProducer : public edm::global::EDProducer<> {
+    
+public:
+  explicit L1HPSPFTauProducer(const edm::ParameterSet&);
+  ~L1HPSPFTauProducer() override {};
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  
+
+private:
+  
+  
+  void produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const override;
+  
+
+  //various needed vars
+  int _nTaus;
+  bool _HW;
+  bool fUseJets_;
+  bool _debug;
+  
+  edm::InputTag srcL1PFCands_;
+  edm::EDGetTokenT<l1t::PFCandidateCollection> tokenL1PFCands_;    
+  //jets
+  edm::InputTag srcL1PFJets_;
+  edm::EDGetTokenT<std::vector<reco::CaloJet>> tokenL1PFJets_;
+  //functions
+  std::vector<l1t::PFTau> processEvent_HW(std::vector<edm::Ptr<l1t::PFCandidate>>& parts, std::vector<edm::Ptr<reco::CaloJet>>&  jets) const;
+
+  static std::vector<L1HPSPFTauEmu::Particle> convertJetsToHW(std::vector<edm::Ptr<reco::CaloJet>>& edmJets); 
+  static std::vector<L1HPSPFTauEmu::Particle> convertEDMToHW(std::vector<edm::Ptr<l1t::PFCandidate>>& edmParticles);
+  static std::vector<l1t::PFTau> convertHWToEDM(std::vector<L1HPSPFTauEmu::Tau> hwTaus);
+  
+};
+
+L1HPSPFTauProducer::L1HPSPFTauProducer(const edm::ParameterSet& cfg)
+   : _nTaus(cfg.getParameter<int>("nTaus")),
+     _HW(cfg.getParameter<bool>("HW")),
+     fUseJets_(cfg.getParameter<bool>("useJets")),
+     _debug(cfg.getParameter<bool>("debug")){ //,
+      srcL1PFCands_ = cfg.getParameter<edm::InputTag>("srcL1PFCands");
+      tokenL1PFCands_ = consumes<l1t::PFCandidateCollection>(srcL1PFCands_);   
+      
+      srcL1PFJets_ = cfg.getParameter<edm::InputTag>("srcL1PFJets");
+      tokenL1PFJets_ = consumes<std::vector<reco::CaloJet>>(srcL1PFJets_); 
+      produces<l1t::PFTauCollection>("HPSTaus");
+      
+      produces<l1t::PFCandidateCollection>("SelPFCands");
```

Comment:
If you capture the return value of `produces()` in `EDPutTokenT` data member, the product instance name strings would not have to be repeated (see example in https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Producing_the_EDProduct)

Reply:
The `l1t::PFCandidateCollection` does not seem to be put in the Event? Please remove the `produces<l1t::PFCandidateCollection>()` call in that case.